prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,470 @@
+ *
+ * This function does the following to handle error:
+ *
+ * 1. write an error cq entry for the operation, if writing
+ *    CQ error entry failed, it will write eq entry.
+ *
+ * 2. increase error counter.
+ *
+ * 3. print warning message with self and peer's raw address
+ *
+ * @param[in]	base_ep     efa_base_ep
+ * @param[in]	ibv_cq_ex   extended ibv cq
+ * @param[in]	err         positive libfabric error code
+ * @param[in]	prov_errno  positive EFA provider specific error code
+ * @param[in]	is_tx       if the error is for TX or RX operation
+ */
+static void efa_cq_handle_error(struct efa_base_ep *base_ep,
+				struct ibv_cq_ex *ibv_cq_ex, int err,
+				int prov_errno, bool is_tx)
+{
+	struct fi_cq_err_entry err_entry;
+	fi_addr_t addr;
+	char err_msg[EFA_ERROR_MSG_BUFFER_LENGTH] = {0};
+	int write_cq_err;
+
+	memset(&err_entry, 0, sizeof(err_entry));
+	efa_cq_construct_cq_entry(ibv_cq_ex, (struct fi_cq_tagged_entry *) &err_entry);
+	err_entry.err = err;
+	err_entry.prov_errno = prov_errno;
+
+	if (is_tx)
+		// TODO: get correct peer addr for TX operation
+		addr = FI_ADDR_NOTAVAIL;
+	else
+		addr = efa_av_reverse_lookup(base_ep->av,
+					     ibv_wc_read_slid(ibv_cq_ex),
+					     ibv_wc_read_src_qp(ibv_cq_ex));
+
+	if (OFI_UNLIKELY(efa_write_error_msg(base_ep, addr, prov_errno,
+					     err_msg,
+					     &err_entry.err_data_size))) {
+		err_entry.err_data_size = 0;
+	} else {
+		err_entry.err_data = err_msg;
+	}
+
+	EFA_WARN(FI_LOG_CQ, "err: %d, message: %s (%d)\n",
+		err_entry.err,
+		err_entry.err_data
+			? (const char *) err_entry.err_data
+			: efa_strerror(err_entry.prov_errno),
+		err_entry.prov_errno);
+
+	efa_show_help(err_entry.prov_errno);
+
+	efa_cntr_report_error(&base_ep->util_ep, err_entry.flags);
+	write_cq_err = ofi_cq_write_error(is_tx ? base_ep->util_ep.tx_cq :
+						  base_ep->util_ep.rx_cq,
+					  &err_entry);
+	if (write_cq_err) {
+		EFA_WARN(
+			FI_LOG_CQ,
+			"Error writing error cq entry when handling %s error\n",
+			is_tx ? "TX" : "RX");
+		efa_base_ep_write_eq_error(base_ep, err, prov_errno);
+	}
+}
+
+/**
+ * @brief handle the event that a TX request has been completed
+ *
+ * @param[in]		base_ep     efa_base_ep
+ * @param[in]		ibv_cq_ex   extended ibv cq
+ * @param[in]		cq_entry    fi_cq_tagged_entry
+ */
+static void efa_cq_handle_tx_completion(struct efa_base_ep *base_ep,
+					struct ibv_cq_ex *ibv_cq_ex,
+					struct fi_cq_tagged_entry *cq_entry)
+{
+	struct util_cq *tx_cq = base_ep->util_ep.tx_cq;
+	int ret = 0;
+
+	/* NULL wr_id means no FI_COMPLETION flag */
+	if (!ibv_cq_ex->wr_id)
+		return;
+
+	/* TX completions should not send peer address to util_cq */
+	if (base_ep->util_ep.caps & FI_SOURCE)
+		ret = ofi_cq_write_src(tx_cq, cq_entry->op_context,
+				       cq_entry->flags, cq_entry->len,
+				       cq_entry->buf, cq_entry->data,
+				       cq_entry->tag, FI_ADDR_NOTAVAIL);
+	else
+		ret = ofi_cq_write(tx_cq, cq_entry->op_context, cq_entry->flags,
```

Comment:
Looks like there are still two copies one after the other in the CQ read path. Can we eliminate one of the copies?

The CQ read path is (correct me if I'm wrong)
* Application calls `fi_cq_read` which we set to `ofi_cq_read`
*  `ofi_cq_read` -> `ofi_cq_readfrom`
    * `ofi_cq_readfrom` first calls `cq->progress(cq)` which we set to `efa_cq_progress` -> `efa_cq_poll_ibv_cq` -> poll IBV CQ -> `ofi_cq_write` and we copy the IBV CQ completion into the Libfabric CQ circular queue
    * `ofi_cq_readfrom` then calls `ofi_cq_read_entries` and we copy the CQ entry from Libfabric CQ circular queue to the application CQ entry buffer

If the progress was more complicated (progress other CQs or EPs for protocol reasons), then we need the circular queue. But in this case, we really don't.

Reply:
Application calls `fi_cq_read` with a `count`. We can stop polling after we get `count` completions. rdma-core is already buffering the CQ entries, so we can just take advantage of that.